### PR TITLE
Add user and home timeline endpoint support

### PR DESCRIPTION
**Pull request by HankG**
Implements Feature Issue #23 

- Has endpoint for the user timeline and home timeline separately
- Fields which came back in testing as null now are optional (may need to be a lot more optional since only pulled about 1000 individual tweets for testing)
- Adds support for pulling the full text back for a Tweet
- Adds ability to pass parameters to the apiUrl builder
- ExtendableApiClient.getInternalListResponse method since lists/arrays of JsonObjects needs to have a ListSerializer path taken by Ktor and it can't resolve that when passed a JsonObject

### CODE REVIEW

File: v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/v1.kt [hunk 1]
```diff
@@ -51,6 +51,18 @@ internal suspend inline fun <reified T> ExtendableApiClient.getInternal(url: Url
     }
 }
 
+@OptIn(Twitter4ktInternalAPI::class)
+internal suspend inline fun <reified T> ExtendableApiClient.getInternalListResponse(url: Url): ApiResult<List<T>> {
```
Comment (yshrsmz) [inline v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/v1.kt:55]: Do we need this?

I think `getInternal` can handle List response too.

File: v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/userTimeline.kt [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+            "max_id" to maxId,
+            "trim_user" to trimUser,
+            "exclude_replies" to excludeReplies,
+            "include_entities" to includeEntities,
+            "include_rts" to includeRts,
+            "tweet_mode" to "extended"
+        )
+    }
+    val url = apiUrl("1.1/statuses/user_timeline.json", parameters = parameters)
+    return apiClient.getInternalListResponse<Tweet>(url.build())
```
Comment (yshrsmz) [inline v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/userTimeline.kt:71]:
```suggestion
    return apiClient.getInternalListResponse(url.build())
```

Kotlin can infer type here(yes, I did this in some other places, and need to remove it)

File: core/core-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/core/Urls.kt [hunk 3]
```diff
@@ -16,12 +16,15 @@
...
+import io.ktor.http.Parameters
 import io.ktor.http.URLBuilder
 
 internal object Urls {
     const val API_ENDPOINT: String = "https://api.twitter.com"
 }
 
-public fun apiUrl(path: String, baseUrl: String = Urls.API_ENDPOINT): URLBuilder {
-    return URLBuilder(baseUrl).path(path)
+public fun apiUrl(path: String, baseUrl: String = Urls.API_ENDPOINT, parameters: Parameters = Parameters.build {}): URLBuilder {
```
Comment (yshrsmz) [inline core/core-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/core/Urls.kt:26]: 👍

File: v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    trimUser: Boolean? = null,
+    excludeReplies: Boolean? = null,
+    includeEntities: Boolean? = null
+): ApiResult<List<Tweet>> {
+    val parameters = Parameters.build {
+        appendNotNulls(
+            "count" to count,
+            "since_id" to sinceId,
+            "max_id" to maxId,
+            "trim_user" to trimUser,
```
Comment (yshrsmz) [inline v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt:56]: Could you remove this as per #21 ?
Sorry for the extra work.

**Review by yshrsmz — CHANGES_REQUESTED** [review]

File: v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/v1.kt (hunk 1 above)
Comment (HankG) [reply v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/v1.kt:55]: > Do we need this?
> 
> I think `getInternal` can handle List response too.

I wasn't able to get it to do that with my experimentation.

File: v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt (hunk 4 above)
Comment (HankG) [reply v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt:56]: Which piece do you want me to remove?

Comment (yshrsmz) [reply v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt:56]: trim_user parameter🙏

File: v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/v1.kt (hunk 1 above)
Comment (yshrsmz) [reply v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/v1.kt:55]: Ok, I'll revisit this later, as I think we don't need to encode to  `JsonObject` in `getInternal` (I have no idea why I wrote this.)

For now, you can keep this implementation 👍

File: v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt (hunk 4 above)
Comment (HankG) [reply v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt]: I believe the latest commit resolves this issue

**Review by yshrsmz — APPROVED** [review]
Thanks!

File: v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+ * @param sinceId Returns results with an ID greater than (that is, more recent than) the specified ID. There are limits to the number of Tweets which can be accessed through the API. If the limit of Tweets has occured since the since_id, the since_id will be forced to the oldest ID available.
+ * @param maxId Returns results with an ID less than (that is, older than) or equal to the specified ID.
+ * @param excludeReplies This parameter will prevent replies from appearing in the returned timeline. Using exclude_replies with the count parameter will mean you will receive up-to count Tweets — this is because the count parameter retrieves that many Tweets before filtering out retweets and replies.
+ * @param includeEntities The entities node will not be included when set to false.
+ */
+@OptIn(Twitter4ktInternalAPI::class)
+public suspend fun StatusesApi.homeTimeline(
+    count: Int? = null,
+    sinceId: Long? = null,
```
Comment (yshrsmz) [inline v1/v1-api/src/commonMain/kotlin/com/codingfeline/twitter4kt/v1/api/statuses/homeTimeline.kt:44]: woops, we should use string id here. will fix these later

**Pull request merged**